peregrine/iqgen/bits/doppler_linear.py

```diff
@@ -0,0 +1,231 @@
+
+    Parameters
+    ----------
+    svTime_s : float
+      Time in seconds at which distance is computed. Please note that  is not
+      a time of the observer.
+
+    Returns
+    -------
+    float
+      Distance to satellite in meters.
+    '''
+    return self.distance0_m + self.speed0_mps * svTime_s + 0.5 * self.acceleration_mps2 * svTime_s * svTime_s
+
+  def computeSpeedMps(self, svTime_s):
+    '''
+    Computes speed of satellite in meters per second.
+
+    Parameters
+    ----------
+    svTime_s : float
+      Time in seconds at which speed is computed. Please note that  is not
+      a time of the observer.
+
+    Returns
+    -------
+    float
+      Speed of satellite in meters per second.
+    '''
+    return self.speed0_mps + self.acceleration_mps2 * svTime_s
+
+  def computeSvTimeS(self, userTime_s):
+    '''
+    Computes SV time from a user time.
+
+    The computation is a solution of the formulae:
+    T_user = T_sv + D(T_sv)/c, for D(t) = D_0 + V*T_sv
+
+    Parameters
+    ----------
+    userTime_s : float
+      Observer's time in seconds
+
+    Returns
+    -------
+    float
+      Satellite vehicle's time at which signal has been generated.
+    '''
+
+    _ax2 = self.acceleration_mps2
+
+    if _ax2 == 0:
+      return userTime_s - self.speed0_mps * userTime_s / scipy.constants.c
+
+    _b = scipy.constants.c + self.speed0_mps
+    _c = self.distance0_m - userTime_s * scipy.constants.c
+
+    _d = _b * _b - 2 * _ax2 * _c
+    _sd = math.sqrt(_d)
+    _r0 = (-_b + _sd) / _ax2
+    _r1 = (-_b - _sd) / _ax2
+    if (_r0 < 0):
+      _res = _r1
+    elif (_r1 < 0):
+      _res = _r0
+    else:
+      _res = min(_r0, _r1)
+
+    return userTime_s - _res  # _tau1
+
+  def computeBatch(self,
+                   userTime0_s,
+                   n_samples,
+                   amplitude,
+                   planFrequency_hz,
+                   ifFrequency_hz,
+                   message,
+                   code):
+    '''
+    Computes signal samples for the doppler object.
+
+    Parameters
+    ----------
+    userTime0_s : float
+      Observer's time in seconds of the beginning of the interval.
+    n_samples : int
+      Number of samples to generate
+    amplitude : float
+      Signal amplitude.
+    planFrequency_hz : float
+      Central carrier frequency in hertz
+    ifFrequency_hz: float
+      Intermediate frequency in hertz
+    message : object
+      Message object for providing access to symbols
+    code : object
+      PRN code object for providing access to chips
+
+    Returns
+    -------
+    signal : numpy.ndarray(n_samples, dtype=float)
+      Generated samples
+    userTimeX_s : float
+      End of interval time in seconds
+    chipAll_idx : numpy.ndarray(n_samples, dtype=float)
+      Code chip phases for the samples
+    chips : numpy.ndarray(n_samples, dtype=int)
+      Code combined with data
+    '''
+
+    deltaUserTime_s = float(n_samples) / if_iface.Chip.SAMPLE_RATE_HZ
```

Comment:
The if_iface.Chip object provides sampling rate to other classes by being passed through? This seems like something that should be passed into the `Doppler` class directly.